log/DESIGN.md

```diff
@@ -0,0 +1,671 @@
+less user friendly (users and bridges would use e.g. a `sync.Pool` to reduce
+the number of heap allocation), less safe (more prone to use after free bugs
+and race conditions), and the benchmark differences were not significant.
+
+### Record.Body as any
+
+[Logs Data Model](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-body)
+defines Body to be `any`.
+We could define `Body` as `any` instead of `string`.
+This could result in a more flexible API
+that could support logging libraries which model log records as any objects.
+
+However, using `any` decreases the performance.[^7]
+
+Additionally, we are not aware of any popular Go logging library
+that uses any objects as log records.
+
+At last, the specification recommends using `string` in Bridge API:
+
+> First-party Applications SHOULD use a string message.
+
+As according to [the specification](https://opentelemetry.io/docs/specs/otel/logs/#new-first-party-application-logs)
+First-party Applications are supposed to use the log bridges
+that integrates with the SDK through Bridge API.
+
+We can always add an additional `StructuredBody any` field in a future release
+if we would need to support structured bodies. This approach would be
+backwards-compatible and should not have such a negative impact on performance
+for the most common scenario where `Body` is a `string`.
+
+### Severity type encapsulating number and text
+
+We could combine severity into a single field defining a type:
+
+```go
+type Severity struct {
+	Number SeverityNumber
+	Text string
+}
+```
+
+However, the [Logs Data Model](https://opentelemetry.io/docs/specs/otel/logs/data-model/#log-and-event-record-definition)
+define it as independent fields.
+It should be more user friendly to have them separated.
+Especially when having getter and setter methods, setting one value
+when the other is already set would be unpleasant.
+
+## Reuse attribute package
+
+It was tempting to reuse the existing
+[https://pkg.go.dev/go.opentelemetry.io/otel/attribute] package
+for defining log attributes and body.
+
+However, this would be wrong because [the log attribute definition](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-attributes)
+is different from [the common attribute definition](https://opentelemetry.io/docs/specs/otel/common/#attribute).
+
+Moreover, it there is nothing telling that [the body definition](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-body)
+has anything in common with a common attribute value.
+
+Therefore, we define new types representing the abstract types defined
+in the [Logs Data Model](https://opentelemetry.io/docs/specs/otel/logs/data-model/#definitions-used-in-this-document).
+
+## Open issues
+
+### Reusing common attributes
+
+Currently, logs attributes are different from common (resource, trace, metrics) attributes.
+However, there may be a desire to make unify them
+as some languages already use common attributes for defining log attributes.[^7]
```

Comment:
I am not sure if it is still an issue given this: https://github.com/open-telemetry/opentelemetry-specification/pull/2888#issuecomment-1904888370.

WDYT?

Reply:
SIG meeting:
We can postpone the implementation of `String` even after making the Bridge API stable as it won't be a breaking change. Especially if we will have no agreement on the `string` representation.